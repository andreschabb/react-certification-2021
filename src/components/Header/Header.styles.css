.ui.menu {
  margin: 0;
}

.ui.menu .item:before {
  position: absolute;
  top: 0;
  right: 0;
  height: 100%;
  width: 0;
}

.clickable {
  cursor: pointer;
}

div.link.icon.item i.icon {
  line-height: 1;
  vertical-align: middle;
  font-size: 2rem;
}

@media only screen and (max-width: 630px) {
  div.right.menu .darkMode {
    display: none !important;
  }
}

@media only screen and (max-width: 430px) {
  div.right.menu .userIcon {
    display: none !important;
  }
}

@media only screen and (max-width: 360px) {
  div.ui.menu .searchInput {
    width: 80%;
  }
}
