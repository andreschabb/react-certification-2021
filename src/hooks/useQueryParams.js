import { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';

const useQueryParams = () => {
  const [searchedText, setSearchedText] = useState('');
  const [videoId, setVideoId] = useState('');
  const location = useLocation();

  useEffect(() => {
    const query = new URLSearchParams(location.search);
    setSearchedText(query.get('searchedText'));
    setVideoId(query.get('id'));
  }, [location]);

  return { searchedText, videoId };
};

export default useQueryParams;
