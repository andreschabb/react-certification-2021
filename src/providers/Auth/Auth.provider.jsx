import React, { useState, useEffect, useContext, useCallback } from 'react';

import { AUTH_STORAGE_KEY } from '../../utils/constants';
import { storage } from '../../utils/storage';

const mockedUser = {
  id: '123',
  name: 'Wizeline',
  avatarUrl:
    'https://media.glassdoor.com/sqll/868055/wizeline-squarelogo-1473976610815.png',
};

const AuthContext = React.createContext(null);

const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error(`Can't use "useAuth" without an AuthProvider!`);
  }
  return context;
};

const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false);
  const [user, setUser] = useState({});

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY);
    const isAuthenticated = Boolean(lastAuthState);

    setUser(lastAuthState);
    setAuthenticated(isAuthenticated);
  }, []);

  const login = useCallback(async (username, password) => {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        if (username === 'wizeline' && password === 'Rocks!') {
          setAuthenticated(true);
          setUser(mockedUser);
          storage.set(AUTH_STORAGE_KEY, mockedUser);
          return resolve(mockedUser);
        }
        return reject(new Error('Username or password invalid'));
      }, 500);
    });
  }, []);

  const logout = useCallback(() => {
    setUser(null);
    setAuthenticated(false);
    storage.set(AUTH_STORAGE_KEY, null);
  }, []);

  return (
    <AuthContext.Provider value={{ login, logout, authenticated, user }}>
      {children}
    </AuthContext.Provider>
  );
};

export { useAuth };
export default AuthProvider;
