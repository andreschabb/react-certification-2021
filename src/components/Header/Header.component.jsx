import './Header.styles.css';
import React, { useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { Checkbox, Icon, Input, Menu, Dropdown } from 'semantic-ui-react';

import { useAuth } from '../../providers/Auth';
import useQueryParams from '../../hooks/useQueryParams';

const Header = ({ onOpenSidebar, onOpenModal, onDarkMode, isDarkMode }) => {
  const history = useHistory();
  const [searchedText, setSearchedText] = useState('');
  const { authenticated, logout } = useAuth();
  const queryParam = useQueryParams().searchedText;

  useEffect(() => {
    setSearchedText(queryParam || '');
  }, [queryParam]);

  const deAuthenticate = () => {
    logout();
    history.push('/');
  };

  const handleKeyDown = (e) => {
    if (e.key === 'Enter') {
      history.push({ pathname: '/results', search: `?searchedText=${searchedText}` });
    }
  };

  return (
    <>
      <Menu data-testid="header-menu">
        <Menu.Item>
          <Icon
            data-testid="header-bars-icon"
            className="clickable"
            name="bars"
            onClick={() => onOpenSidebar(true)}
          />
        </Menu.Item>
        <Menu.Item className="searchInput">
          <Input
            data-testid="header-search"
            icon="search"
            placeholder="Search..."
            value={searchedText}
            onKeyDown={handleKeyDown}
            onChange={(event) => setSearchedText(event.target.value)}
          />
        </Menu.Item>

        <Menu.Menu position="right">
          <Menu.Item className="darkMode">
            <Checkbox
              data-testid="header-dark-mode"
              label="Dark Mode"
              slider
              onClick={() => onDarkMode(!isDarkMode)}
            />
          </Menu.Item>
          <Dropdown
            data-testid="header-dropdown"
            icon={isDarkMode ? 'user circle' : 'user circle outline'}
            className="link icon item userIcon"
          >
            <Dropdown.Menu>
              {!authenticated ? (
                <Dropdown.Item
                  data-testid="header-dropdown-login"
                  onClick={() => onOpenModal(true)}
                >
                  Login
                </Dropdown.Item>
              ) : (
                <Dropdown.Item
                  data-testid="header-dropdown-logout"
                  onClick={deAuthenticate}
                >
                  Logout
                </Dropdown.Item>
              )}
            </Dropdown.Menu>
          </Dropdown>
        </Menu.Menu>
      </Menu>
    </>
  );
};

export default Header;
