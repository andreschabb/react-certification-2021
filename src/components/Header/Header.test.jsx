import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import { useAuth } from '../../providers/Auth';
import useQueryParams from '../../hooks/useQueryParams';
import Header from './Header.component';

const mockHistoryPush = jest.fn();
const mockLogout = jest.fn();

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

jest.mock('../../providers/Auth');

jest.mock('../../hooks/useQueryParams');

describe('Header', () => {
  beforeEach(() => {
    useAuth.mockImplementation(() => ({ authenticated: false }));
    useQueryParams.mockImplementation(() => ({}));
  });

  it('should renders', () => {
    const { container } = render(<Header />);

    expect(screen.getByTestId('header-menu')).toBeTruthy();

    expect(container).toMatchSnapshot();
  });

  it('should opens sidebar', () => {
    const onOpenSidebar = jest.fn();

    render(<Header onOpenSidebar={onOpenSidebar} />);

    fireEvent.click(screen.getByTestId('header-bars-icon'));

    expect(onOpenSidebar).toHaveBeenCalledTimes(1);
  });

  it('should changes the theme mode', () => {
    const onDarkMode = jest.fn();

    render(<Header onDarkMode={onDarkMode} />);

    fireEvent.click(screen.getByTestId('header-dark-mode'));

    expect(onDarkMode).toHaveBeenCalledTimes(1);
  });

  it('should changes the icon with dark mode', () => {
    render(<Header isDarkMode />);

    let userIcon = document.body.getElementsByClassName('user circle outline');
    expect(userIcon[0]).toBeFalsy();
    userIcon = document.body.getElementsByClassName('user circle');
    expect(userIcon[0]).toBeTruthy();
  });

  it('should changes the icon with light mode', () => {
    render(<Header />);

    const userIcon = document.body.getElementsByClassName('user circle outline');
    expect(userIcon[0]).toBeTruthy();
  });

  it('should opens login modal', () => {
    const onOpenModal = jest.fn();

    render(<Header onOpenModal={onOpenModal} />);

    fireEvent.click(screen.getByTestId('header-dropdown'));
    fireEvent.click(screen.getByTestId('header-dropdown-login'));

    expect(onOpenModal).toHaveBeenCalledTimes(1);
  });

  it('should allows user to logout', () => {
    useAuth.mockImplementation(() => ({ authenticated: true, logout: mockLogout }));
    render(<Header />);

    fireEvent.click(screen.getByTestId('header-dropdown'));
    fireEvent.click(screen.getByTestId('header-dropdown-logout'));
    expect(mockLogout).toHaveBeenCalledTimes(1);
    expect(mockHistoryPush).toHaveBeenLastCalledWith('/');
  });

  it('should redirect when user press enter on input search', () => {
    const expected = {
      pathname: '/results',
      search: '?searchedText=Wizeline',
    };
    render(<Header />);

    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 'Wizeline' } });
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });
    expect(mockHistoryPush).toHaveBeenCalledTimes(1);
    expect(mockHistoryPush).toHaveBeenLastCalledWith(expected);
  });

  it('should not redirect when user press any other key on input search', () => {
    render(<Header />);

    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 'Wizeline' } });
    fireEvent.keyDown(input, { key: 'Space', code: 'Space' });
    expect(mockHistoryPush).toHaveBeenCalledTimes(0);
  });
});
