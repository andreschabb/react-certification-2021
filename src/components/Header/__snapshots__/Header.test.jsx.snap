// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should renders 1`] = `
<div>
  <div
    class="ui menu"
    data-testid="header-menu"
  >
    <div
      class="item"
    >
      <i
        aria-hidden="true"
        class="bars icon clickable"
        data-testid="header-bars-icon"
      />
    </div>
    <div
      class="item searchInput"
    >
      <div
        class="ui icon input"
        data-testid="header-search"
      >
        <input
          placeholder="Search..."
          type="text"
          value=""
        />
        <i
          aria-hidden="true"
          class="search icon"
        />
      </div>
    </div>
    <div
      class="right menu"
    >
      <div
        class="item darkMode"
      >
        <div
          class="ui slider checkbox"
          data-testid="header-dark-mode"
        >
          <input
            class="hidden"
            readonly=""
            tabindex="0"
            type="checkbox"
            value=""
          />
          <label>
            Dark Mode
          </label>
        </div>
      </div>
      <div
        aria-expanded="false"
        class="ui dropdown link icon item userIcon"
        data-testid="header-dropdown"
        role="listbox"
        tabindex="0"
      >
        <i
          aria-hidden="true"
          class="user circle outline icon"
        />
        <div
          class="menu transition"
        >
          <div
            class="item"
            data-testid="header-dropdown-login"
            role="option"
          >
            Login
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
